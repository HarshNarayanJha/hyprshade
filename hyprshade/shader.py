from __future__ import annotations

from glob import iglob
from os import path

from more_itertools import first

from . import hyprctl
from .constants import SHADER_DIRS


def _stripped_basename(s: str) -> str:
    return path.splitext(path.basename(s))[0]


class Shader:
    def __init__(self, shader_name_or_path: str):
        self._given_path = (
            shader_name_or_path
            if path.dirname(shader_name_or_path) and path.isfile(shader_name_or_path)
            else None
        )
        self._name = _stripped_basename(shader_name_or_path)

    def __str__(self) -> str:
        return self._name

    def __repr__(self) -> str:
        return f"Shader({self._name!r})"

    @property
    def name(self) -> str:
        return self._name

    def on(self) -> None:
        path_ = self._resolve_path()
        hyprctl.set_screen_shader(path_)

    @staticmethod
    def off() -> None:
        hyprctl.clear_screen_shader()

    @staticmethod
    def current() -> Shader | None:
        name = hyprctl.get_screen_shader()
        return None if name is None else Shader(name)

    def samefile(self, __value: Shader) -> bool:
        s, s2 = self._resolve_path(), __value._resolve_path()
        return path.samefile(s, s2)

    def _resolve_path(self) -> str:
        if self._given_path:
            return self._given_path

        for dir in SHADER_DIRS:
            path_ = first(iglob(f"{self._name}*", root_dir=dir), None)
            if path_ is not None:
                return path.join(dir, path_)

        raise FileNotFoundError(f"Shader {self._name} does not exist")
