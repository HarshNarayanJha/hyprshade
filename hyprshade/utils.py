import os
from os import path


def xdg_config_home():
    config_home = os.getenv("XDG_CONFIG_HOME")
    if config_home is None:
        home = os.getenv("HOME")
        if home is None:
            raise ValueError("$HOME environment variable is not set")
        config_home = path.join(home, ".config")
    return config_home
