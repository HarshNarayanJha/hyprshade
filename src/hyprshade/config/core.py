from __future__ import annotations

import os
import tomllib

from more_itertools import first_true

from .model import RootConfig
from .utils import hypr_config_home, hyprshade_config_home


class Config:
    model: RootConfig

    def __init__(self, path: str | None = None):
        path = path or Config._get_path()
        if path is None:
            raise FileNotFoundError(
                "Could not find a config file; see https://github.com/loqusion/hyprshade#scheduling"
            )
        self.model = RootConfig(Config._load(path), path=path)

    @staticmethod
    def _load(path: str) -> dict:
        with open(path, "rb") as f:
            return tomllib.load(f)

    @staticmethod
    def _get_path() -> str | None:
        candidates = [
            os.getenv("HYPRSHADE_CONFIG"),
            os.path.join(hypr_config_home(), "hyprshade.toml"),
            os.path.join(hyprshade_config_home(), "config.toml"),
        ]
        return first_true((c for c in candidates if c is not None), pred=os.path.isfile)
