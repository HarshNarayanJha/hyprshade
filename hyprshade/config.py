import copy
from datetime import time
from os import path
from typing import Any, TypeVar, cast

import tomllib

from hyprshade.utils import hyprshade_config_home, is_time_between

K = TypeVar("K")
V = TypeVar("V")


class ShadeConfig:
    name: str
    default: bool
    start_time: time | None = None
    end_time: time | None = None

    def __init__(self, shade_config: dict):
        try:
            self._shade_config = shade_config
            self.name = shade_config["name"]
            self.default = shade_config.get("default", False)
            if not self.default:
                self.start_time = shade_config["start_time"]
                self.end_time = shade_config.get("end_time")
        except KeyError as e:
            raise ValueError(f"Missing key {e} in shade config") from e

    def __repr__(self) -> str:
        return (
            f"ShadeConfig(name={self.name}, default={self.default}, "
            f"start_time={self.start_time}, end_time={self.end_time})"
        )


class Config:
    shades: list[ShadeConfig]
    default: ShadeConfig | None = None

    def __init__(self, config_path: str | None = None):
        if config_path is None:
            config_path = Config.get_path()
        config = Config.load(config_path)
        shades_list: list = config["shades"]
        self.shades = list(map(ShadeConfig, shades_list))
        self._coerce()

    def _coerce(self):
        """Performs config file validation and deduplicates time entries.

        If a shade is marked as default, it is set as the default shade.
        If a shade has an end time in the time range of another shade,
        the end time is removed.

        If more than one shade is marked as default, or if there are multiple
        shades with the same start time, an error is raised.
        """

        start_times: dict[time, ShadeConfig] = {}
        shades_mut = copy.deepcopy(self.shades)
        for shade, shade_mut in zip(self.shades, shades_mut, strict=True):
            if shade.default:
                self._set_default(shade)
                continue

            if shade.start_time in start_times:
                raise ValueError(
                    f"Multiple shades with start time {shade.start_time}: "
                    f"'{shade.name}' and '{start_times[shade.start_time].name}'"
                )

            if shade.start_time is not None:
                start_times[shade.start_time] = shade_mut

            if shade.end_time is not None:

                def other_shade_has_time_range(s: ShadeConfig) -> bool:
                    return (
                        s.start_time is not None
                        and s.end_time is not None
                        and s.name != shade.name  # noqa: B023
                    )

                if shade.start_time is not None:
                    for cshade in filter(other_shade_has_time_range, self.shades):
                        if is_time_between(
                            cast(time, cshade.end_time),
                            shade.start_time,
                            shade.end_time,
                        ):
                            cshade_mut = next(
                                s for s in shades_mut if s.name == cshade.name
                            )
                            cshade_mut.end_time = None
                            break

                for cshade in filter(other_shade_has_time_range, self.shades):
                    if is_time_between(
                        shade.end_time,
                        cast(time, cshade.start_time),
                        cast(time, cshade.end_time),
                    ):
                        shade_mut.end_time = None
                        break

        self.shades = shades_mut

    @staticmethod
    def load(config_path: str) -> dict[str, Any]:
        with open(config_path, "rb") as f:
            return tomllib.load(f)

    @staticmethod
    def get_path() -> str:
        return path.join(hyprshade_config_home(), "config.toml")

    def _set_default(self, shade: ShadeConfig):
        if self.default is not None:
            raise ValueError(
                f"Multiple default shades: '{self.default.name}' and '{shade.name}'\n"
                "Please set only one shade as default."
            )
        self.default = shade
