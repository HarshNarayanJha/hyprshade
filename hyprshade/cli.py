import os
import sys
from datetime import datetime
from functools import cache
from itertools import chain
from os import path

import click

from .constants import SHADER_DIRS
from .helpers import resolve_shader_path
from .hyprctl import clear_screen_shader, get_screen_shader, set_screen_shader
from .utils import systemd_user_config_home


@click.group()
def cli():
    pass


def main():
    try:
        cli()
    except Exception as e:
        click.echo(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


@cli.command()
@click.argument("shader_name_or_path")
def on(shader_name_or_path: str):
    """Turn on screen shader."""

    shader_path = resolve_shader_path(shader_name_or_path)
    set_screen_shader(shader_path)


@cli.command()
def off():
    """Turn off screen shader."""

    clear_screen_shader()


@cli.command()
@click.argument("shader_name_or_path", default=None)
@click.option(
    "--fallback",
    metavar="SHADER",
    help="Shader to switch to instead of toggling off.",
)
@click.option(
    "--fallback-default",
    is_flag=True,
    default=False,
    help="Use default shader as fallback. (see --fallback)",
)
@click.pass_context
def toggle(
    ctx: click.Context,
    shader_name_or_path: str | None,
    fallback: str | None,
    fallback_default: bool,
):
    """Toggle screen shader.

    If run with no arguments, SHADER_NAME_OR_PATH is inferred based on schedule.

    When --fallback is specified, will toggle between SHADER_NAME_OR_PATH and the
    fallback shader. --fallback-default will toggle between SHADER_NAME_OR_PATH and the
    default shader.
    """

    from .config import Schedule

    @cache
    def schedule() -> Schedule:
        from .config import Config

        try:
            return Config().to_schedule()
        except FileNotFoundError as e:
            click.echo(f"Error: {e}")
            sys.exit(1)

    if fallback and fallback_default:
        raise click.BadOptionUsage(
            "--fallback", "Cannot specify both --fallback and --fallback-default"
        )

    t = datetime.now().time()

    if fallback_default:
        fallback = schedule().default_shade_name

    def toggle_off():
        if fallback is None:
            ctx.invoke(off)
            ctx.exit()
        else:
            ctx.invoke(on, shader_name_or_path=fallback)
            ctx.exit()

    shade = shader_name_or_path or schedule().find_shade(t)
    if shade is None:
        ctx.invoke(off)
        ctx.exit()
    shade = resolve_shader_path(shade)

    current_shader = get_screen_shader()
    if current_shader is not None and path.samefile(shade, current_shader):
        toggle_off()
        ctx.exit()

    ctx.invoke(on, shader_name_or_path=shade)


@cli.command()
@click.pass_context
def auto(ctx):
    """Turn on/off screen shader based on schedule."""

    from .config import Config

    t = datetime.now().time()
    shade = Config().to_schedule().find_shade(t)

    if shade is not None:
        ctx.invoke(on, shade_or_path=shade)
        ctx.exit()
    ctx.invoke(off)


@cli.command()
def install():
    """Install systemd user units."""

    from .config import Config

    schedule = Config().to_schedule()

    with open(path.join(systemd_user_config_home(), "hyprshade.service"), "w") as f:
        f.write(
            """[Unit]
Description=Apply screen filter

[Service]
Type=oneshot
ExecStart="/usr/bin/hyprshade" auto"""
        )

    with open(path.join(systemd_user_config_home(), "hyprshade.timer"), "w") as f:
        on_calendar = "\n".join(
            sorted([f"OnCalendar=*-*-* {x}" for x in schedule.on_calendar_entries()])
        )
        f.write(
            f"""[Unit]
Description=Apply screen filter on schedule

[Timer]
{on_calendar}

[Install]
WantedBy=timers.target"""
        )


@cli.command()
def ls():
    """List available screen shaders."""

    current_shader = get_screen_shader()
    shader_base = path.basename(current_shader) if current_shader else None

    for shader in chain(
        *map(
            os.listdir,
            SHADER_DIRS,
        )
    ):
        c = "*" if shader == shader_base else " "
        shader, _ = path.splitext(shader)
        click.echo(f"{c} {shader}")
