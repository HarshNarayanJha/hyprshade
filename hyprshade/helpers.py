from functools import cache
from glob import iglob
from os import path
from typing import Literal

from more_itertools import first

from hyprshade.utils import systemd_user_config_home

from .config import Schedule
from .constants import SHADER_DIRS


def resolve_shader_path(shader_name_or_path: str) -> str:
    if path.isfile(shader_name_or_path):
        return shader_name_or_path

    for shaders_dir in SHADER_DIRS:
        shader_path = first(
            iglob(f"{shader_name_or_path}*", root_dir=shaders_dir), None
        )
        if shader_path is not None:
            return path.join(shaders_dir, shader_path)

    raise FileNotFoundError(f"Shader {shader_name_or_path} does not exist")


@cache
def schedule_from_config(config_path: str | None = None) -> Schedule:
    from .config import Config

    return Config(config_path).to_schedule()


SystemdUnitType = Literal["service", "timer"]


def write_systemd_user_unit(unit_type: SystemdUnitType, body: str):
    with open(
        path.join(systemd_user_config_home(), f"hyprshade.{unit_type}"), "w"
    ) as f:
        f.write(body)
