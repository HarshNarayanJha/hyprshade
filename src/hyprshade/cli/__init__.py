from __future__ import annotations

import logging
from typing import Final

import click

from hyprshade.cli.utils import ContextObject
from hyprshade.config.core import Config

from .auto import auto
from .current import current
from .install import install
from .ls import ls
from .off import off
from .on import on
from .toggle import toggle

COMMANDS: Final = [
    auto,
    current,
    install,
    ls,
    off,
    on,
    toggle,
]


@click.group()
@click.version_option()
@click.option("-v", "--verbose", is_flag=True, help="Enable verbose output")
@click.pass_context
def cli(ctx: click.Context, verbose: bool):
    level = logging.DEBUG if verbose else logging.WARNING
    logging.basicConfig(level=level)

    try:
        config = Config()
    except FileNotFoundError:
        config = None
    ctx.obj = ContextObject(config)


for command in COMMANDS:
    cli.add_command(command)


def main():
    try:
        return cli()
    except Exception as e:
        if logging.getLogger().getEffectiveLevel() <= logging.DEBUG:
            raise e
        message = str(e) or repr(e)
        click.echo(f"{click.style('Error', fg='red')}: {message}", err=True)
        click.secho("Use --verbose to see the full traceback", fg="yellow", err=True)
        return 1
