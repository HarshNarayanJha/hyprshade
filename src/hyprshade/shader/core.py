from __future__ import annotations

import logging
import os
from collections.abc import Callable
from functools import cached_property
from typing import Any, Final, TypeVar, Union

from more_itertools import flatten

from hyprshade.template import mustache
from hyprshade.template.constants import TEMPLATE_EXTENSIONS
from hyprshade.utils.fs import scandir_recursive
from hyprshade.utils.path import strip_all_extensions, stripped_basename
from hyprshade.utils.xdg import user_state_dir

from . import hyprctl
from .dirs import ShaderDirs

T = TypeVar("T")
PossiblyLazy = Union[T, Callable[[], T]]  # noqa: UP007

ShaderVariables = dict[str, Any]


class PureShader:
    _name: str
    _given_path: str | None
    _template_instance_path: str | None

    def __init__(
        self, shader_name_or_path: str, *, template_instance_path: str | None = None
    ):
        if shader_name_or_path.find(os.path.sep) != -1:
            self._name = PureShader.path_to_name(shader_name_or_path)
            self._given_path = os.path.abspath(shader_name_or_path)
        else:
            if shader_name_or_path.find(".") != -1:
                raise ValueError(
                    f"Shader name '{shader_name_or_path}' must not contain a '.' character"
                )
            self._name = shader_name_or_path
            self._given_path = None
        self._template_instance_path = template_instance_path

    def __eq__(self, __value: object) -> bool:
        if not isinstance(__value, PureShader):
            return False
        try:
            s1, s2 = self._resolve_path(), __value._resolve_path()
        except FileNotFoundError:
            return False
        return os.path.samefile(s1, s2)

    def __str__(self) -> str:
        return self._name

    def __repr__(self) -> str:
        return f"{type(self).__name__}({self._name!r})"

    @property
    def name(self) -> str:
        return self._name

    def path(self) -> str:
        return self._resolve_path()

    @property
    def template_instance_path(self) -> str | None:
        return self._template_instance_path

    @staticmethod
    def path_to_name(path: str) -> str:
        return stripped_basename(path)

    def _resolve_path(self) -> str:
        if self._given_path:
            if not os.path.exists(self._given_path):
                raise FileNotFoundError(f"No file found at '{self._given_path}'")
            return self._given_path
        return self._resolve_path_from_shader_dirs()

    def _resolve_path_from_shader_dirs(self) -> str:
        dirs = Shader.dirs.all()
        all_files = flatten(scandir_recursive(d, max_depth=5) for d in dirs)
        for file in all_files:
            if strip_all_extensions(file.name) == self._name:
                return file.path

        raise FileNotFoundError(
            f"Shader '{self._name}' could not be found in any of the following"
            " directories:\n\t"
            "{}".format("\n\t".join(dirs))
        )


class Shader(PureShader):
    dirs: Final = ShaderDirs
    _variables: PossiblyLazy[ShaderVariables | None]

    TEMPLATE_METADATA_PREFIX: Final = "// SOURCE:"

    def __init__(
        self,
        shader_name_or_path: str,
        variables: PossiblyLazy[ShaderVariables | None],
        template_instance_path: str | None = None,
    ):
        super().__init__(
            shader_name_or_path, template_instance_path=template_instance_path
        )
        self._variables = variables

    def on(self) -> None:
        source_path = self._resolve_path()
        _, source_path_extension = os.path.splitext(os.path.basename(source_path))
        if source_path_extension.strip(".") in TEMPLATE_EXTENSIONS:
            rendered_path = self._render_template(source_path)
        else:
            rendered_path = source_path
        logging.debug(f"Turning on shader '{self._name}' at '{rendered_path}'")
        hyprctl.set_screen_shader(rendered_path)

    @staticmethod
    def off() -> None:
        hyprctl.clear_screen_shader()

    @staticmethod
    def current() -> PureShader | None:
        path = hyprctl.get_screen_shader()
        if path is not None and (
            os.path.commonpath([path, user_state_dir("hyprshade")])
            == user_state_dir("hyprshade")
        ):
            return PureShader(
                Shader._extract_source_path_from_template(path),
                template_instance_path=path,
            )
        return None if path is None else PureShader(path)

    @cached_property
    def variables(self) -> ShaderVariables | None:
        if callable(self._variables):
            return self._variables()
        return self._variables

    def _render_template(self, path: str) -> str:
        with open(path) as f:
            content = mustache.render(f, self.variables)
        out_path = Shader._template_path_from_source_path(path)
        os.makedirs(os.path.dirname(out_path), exist_ok=True)
        with open(out_path, "w") as f:
            f.write(f"{Shader.TEMPLATE_METADATA_PREFIX} {path}\n")
            f.write(content)
        return out_path

    @staticmethod
    def _template_path_from_source_path(path: str) -> str:
        file_name, _ = os.path.splitext(os.path.basename(path))
        return os.path.join(user_state_dir("hyprshade"), file_name)

    @staticmethod
    def _extract_source_path_from_template(path: str) -> str:
        with open(path) as f:
            first_line = f.readline()
            if not first_line.startswith(Shader.TEMPLATE_METADATA_PREFIX):
                raise ValueError(
                    f"Expected first line of file '{path}' to start with '{Shader.TEMPLATE_METADATA_PREFIX}'"
                )
            source_path = first_line.removeprefix(
                Shader.TEMPLATE_METADATA_PREFIX
            ).strip()
            return source_path
