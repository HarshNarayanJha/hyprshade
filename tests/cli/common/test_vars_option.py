import pytest
from click.testing import CliRunner
from syrupy.assertion import SnapshotAssertion

from hyprshade.cli import cli
from hyprshade.shader import hyprctl
from hyprshade.shader.core import Shader
from tests.types import ConfigFactory, ShaderPathFactory

pytestmark = [
    pytest.mark.requires_hyprland(),
]


@pytest.mark.parametrize("subcommand", ["on", "toggle"])
class TestVarOption:
    def test(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=("""const int key = {{key}};\n""" """void main() {}"""),
        )
        result = runner.invoke(cli, [subcommand, "shader", "--var", "key=3"])

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_multiple(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=(
                """const int WORLD = {{world}};\n"""
                """const int key = {{key}};\n"""
                """void main() {}"""
            ),
        )
        result = runner.invoke(
            cli, [subcommand, "shader", "--var", "key=world", "--var", "world=5"]
        )

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_dot(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=("""const int bar = {{foo.bar}};\n""" """void main() {}"""),
        )
        result = runner.invoke(cli, [subcommand, "shader", "--var", "foo.bar=3"])

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_dot_shallow_merge(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=(
                """const int bar = {{foo.bar}};\n"""
                """const int baz = {{foo.baz}};\n"""
                """const int qux = {{foo.qux}};\n"""
                """void main() {}"""
            ),
        )
        result = runner.invoke(
            cli,
            [
                subcommand,
                "shader",
                "--var",
                "foo.bar=1",
                "--var",
                "foo.baz=2",
                "--var",
                "foo.qux=3",
            ],
        )

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_dot_deep_merge(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=(
                """const int bar = {{foo.bar}};\n"""
                """const int baz_qux = {{foo.baz.qux}};\n"""
                """const int baz_baz_qux = {{foo.baz.baz.qux}};\n"""
                """void main() {}"""
            ),
        )
        result = runner.invoke(
            cli,
            [
                subcommand,
                "shader",
                "--var",
                "foo.bar=1",
                "--var",
                "foo.baz.qux=2",
                "--var",
                "foo.baz.baz.qux=3",
            ],
        )

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_override(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=(
                """const int WORLD = {{world}};\n"""
                """const int key = {{key}};\n"""
                """void main() {}"""
            ),
        )
        result = runner.invoke(
            cli,
            [
                subcommand,
                "shader",
                "--var",
                "key=planet",
                "--var",
                "key=world",
                "--var",
                "world=5",
            ],
        )

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_override_deep(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=(
                """const int foo_bar_baz = {{foo.bar.baz}};\n""" """void main() {}"""
            ),
        )
        result = runner.invoke(
            cli,
            [
                subcommand,
                "shader",
                "--var",
                "foo.bar=0",
                "--var",
                "foo.bar.baz=1",
            ],
        )

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_override_from_deep(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=("""const int foo_bar = {{foo.bar}};\n""" """void main() {}"""),
        )
        result = runner.invoke(
            cli,
            [
                subcommand,
                "shader",
                "--var",
                "foo.bar.baz=0",
                "--var",
                "foo.bar=1",
            ],
        )

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_merge_with_config(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
        config_factory: ConfigFactory,
        snapshot: SnapshotAssertion,
    ):
        shader_path_factory(
            "shader",
            extension="glsl.mustache",
            text=(
                """const int r = {{balance.r}};\n"""
                """const int g = {{balance.g}};\n"""
                """const int b = {{balance.b}};\n"""
                """void main() {}"""
            ),
        )
        config_factory.write(
            {
                "shaders": [
                    {
                        "name": "shader",
                        "config": {"balance": {"r": 1, "g": 5}},
                    }
                ]
            }
        )
        result = runner.invoke(
            cli,
            [
                subcommand,
                "shader",
                "--var",
                "balance.g=2",
                "--var",
                "balance.b=3",
            ],
        )

        assert result.exit_code == 0
        current_shader_path = hyprctl.get_screen_shader()
        assert current_shader_path is not None
        assert (
            Shader._get_template_instance_content_without_metadata(current_shader_path)
            == snapshot
        )

    def test_no_equals(
        self,
        subcommand: str,
        runner: CliRunner,
        shader_path_factory: ShaderPathFactory,
    ):
        shader_path_factory(
            "shader", extension="glsl.mustache", text="""void main() {}"""
        )
        result = runner.invoke(cli, [subcommand, "shader", "--var", "key"])

        assert result.exit_code != 0
        assert "Invalid value for '--var' / '-V'" in result.stderr
