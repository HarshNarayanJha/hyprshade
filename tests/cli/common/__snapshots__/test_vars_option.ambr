# serializer version: 1
# name: TestVarOption.test[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int key = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int key = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_dot[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int bar = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_dot[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int bar = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_dot_deep_merge[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int bar = 1;
  const int baz_qux = 2;
  const int baz_baz_qux = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_dot_deep_merge[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int bar = 1;
  const int baz_qux = 2;
  const int baz_baz_qux = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_dot_shallow_merge[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int bar = 1;
  const int baz = 2;
  const int qux = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_dot_shallow_merge[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int bar = 1;
  const int baz = 2;
  const int qux = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_merge_with_config[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int r = 1;
  const int g = 2;
  const int b = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_merge_with_config[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int r = 1;
  const int g = 2;
  const int b = 3;
  void main() {}
  '''
# ---
# name: TestVarOption.test_multiple[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int WORLD = 5;
  const int key = WORLD;
  void main() {}
  '''
# ---
# name: TestVarOption.test_multiple[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int WORLD = 5;
  const int key = WORLD;
  void main() {}
  '''
# ---
# name: TestVarOption.test_override[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int WORLD = 5;
  const int key = WORLD;
  void main() {}
  '''
# ---
# name: TestVarOption.test_override[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int WORLD = 5;
  const int key = WORLD;
  void main() {}
  '''
# ---
# name: TestVarOption.test_override_deep[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int foo_bar_baz = 1;
  void main() {}
  '''
# ---
# name: TestVarOption.test_override_deep[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int foo_bar_baz = 1;
  void main() {}
  '''
# ---
# name: TestVarOption.test_override_from_deep[on]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int foo_bar = 1;
  void main() {}
  '''
# ---
# name: TestVarOption.test_override_from_deep[toggle]
  '''
  // This file was generated by Hyprshade.
  // Do not edit it directly.
  
  const int foo_bar = 1;
  void main() {}
  '''
# ---
